src/sources/apache_metrics/mod.rs

```diff
@@ -0,0 +1,394 @@
+use crate::{
+    config::{self, GlobalOptions},
+    event::metric::{Metric, MetricKind, MetricValue},
+    internal_events::{
+        ApacheMetricsErrorResponse, ApacheMetricsEventReceived, ApacheMetricsHttpError,
+        ApacheMetricsParseError, ApacheMetricsRequestCompleted,
+    },
+    shutdown::ShutdownSignal,
+    Event, Pipeline,
+};
+use chrono::Utc;
+use futures::{
+    compat::{Future01CompatExt, Sink01CompatExt},
+    future, stream, FutureExt, StreamExt, TryFutureExt,
+};
+use futures01::Sink;
+use hyper::{Body, Client, Request};
+use hyper_openssl::HttpsConnector;
+use serde::{Deserialize, Serialize};
+use snafu::ResultExt;
+use std::collections::BTreeMap;
+use std::time::{Duration, Instant};
+
+mod parser;
+
+#[derive(Deserialize, Serialize, Clone, Debug)]
+struct ApacheMetricsConfig {
+    endpoints: Vec<String>,
+    #[serde(default = "default_scrape_interval_secs")]
+    scrape_interval_secs: u64,
+    #[serde(default = "default_namespace")]
+    namespace: String,
+}
+
+pub fn default_scrape_interval_secs() -> u64 {
+    15
+}
+
+pub fn default_namespace() -> String {
+    "apache".to_string()
+}
+
+#[typetag::serde(name = "apache_metrics")]
+impl crate::config::SourceConfig for ApacheMetricsConfig {
+    fn build(
+        &self,
+        _name: &str,
+        _globals: &GlobalOptions,
+        shutdown: ShutdownSignal,
+        out: Pipeline,
+    ) -> crate::Result<super::Source> {
+        let urls = self
+            .endpoints
+            .iter()
+            .map(|endpoint| endpoint.parse::<http::Uri>())
+            .collect::<Result<Vec<_>, _>>()
+            .context(super::UriParseError)?;
+
+        Ok(apache_metrics(
+            urls,
+            self.scrape_interval_secs,
+            self.namespace.clone(),
+            shutdown,
+            out,
+        ))
+    }
+
+    fn output_type(&self) -> crate::config::DataType {
+        config::DataType::Metric
+    }
+
+    fn source_type(&self) -> &'static str {
+        "apache_metrics"
+    }
+}
+
+fn apache_metrics(
+    urls: Vec<http::Uri>,
+    interval: u64,
+    namespace: String,
+    shutdown: ShutdownSignal,
+    out: Pipeline,
+) -> super::Source {
+    let out = out
+        .sink_map_err(|e| error!("error sending metric: {:?}", e))
+        .sink_compat();
+    let task = tokio::time::interval(Duration::from_secs(interval))
+        .take_until(shutdown.compat())
+        .map(move |_| stream::iter(urls.clone())) // TODO remove clone?
```

Comment:
I'd like to get rid of this clone, but was unable to figure out how without annoying the borrow checker.